File: Cpp/Add-Binary.cpp
#include <string>

class Solution {
public:
    /**
     * @param a a number
     * @param b a number
     * @return the result
     */
    string addBinary(string& a, string& b) {
        // Write your code here
        int d = 0, index1 = a.length() - 1, index2 = b.length() - 1;
        string str = "";
        char ch;
        while (index1 >= 0 && index2 >= 0) {
            int temp = a[index1] - '0' + b[index2] - '0' + d;
            ch = temp % 2 + '0';
            d = temp / 2;
            str += ch;
            index1 --;
            index2 --;
        }
        while (index1 >= 0) {
            int temp = a[index1] - '0' + d;
            ch = temp % 2 + '0';
            d = temp / 2;
            str += ch;
            index1 --;
        }
        while (index2 >= 0) {
            int temp = b[index2] - '0' + d;
            ch = temp % 2 + '0';
            d = temp / 2;
            str += ch;
            index2 --;
        }
        if (d) {
            ch = d + '0';
            str += ch;
        }
        reverse(str.begin(), str.end());
        return str;
    }
};

File: Python/Add-Binary.py
class Solution:
    # @param {string} a a number
    # @param {string} b a number
    # @return {string} the result
    def addBinary(self, a, b):
        # Write your code here
        pa = len(a) - 1
        pb = len(b) - 1
        s = ""
        d = 0
        while pa >= 0 and pb >= 0:
            cur = d + int(a[pa]) + int(b[pb])
            pa -= 1
            pb -= 1
            s += str(cur % 2)
            d = cur / 2
        while pa >= 0:
            cur = d + int(a[pa])
            pa -= 1
            s += str(cur % 2)
            d = cur / 2
        while pb >= 0:
            cur = d + int(b[pb])
            pb -= 1
            s += str(cur % 2)
            d = cur / 2
        if d > 0:
            s += str(d)
        rs = ""
        for i in range(len(s)):
            rs += s[len(s) - 1 - i]
        return rs